k-distribution/src/main/scripts/lib/pyk/pyk/subst.py

```diff
@@ -0,0 +1,113 @@
+from dataclasses import dataclass
+from itertools import chain
+from typing import Dict, Iterator, Mapping, Optional, TypeVar
+
+from .kast import KApply, KInner, KRewrite, KSequence, KToken, KVariable
+from .kastManip import bottom_up
+from .utils import FrozenDict
+
+K = TypeVar('K')
+V = TypeVar('V')
+
+
+@dataclass(frozen=True)
+class Subst(Mapping[str, KInner]):
+    subst: FrozenDict[str, KInner]
+
+    def __init__(self, subst: Mapping[str, KInner] = {}):
+        object.__setattr__(self, 'subst', FrozenDict({k: v for k, v in subst.items() if type(v) is not KVariable or v.name != k}))
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self.subst)
+
+    def __len__(self) -> int:
+        return len(self.subst)
+
+    def __getitem__(self, key: str) -> KInner:
+        return self.subst[key]
+
+    def __mul__(self, other: 'Subst') -> 'Subst':
+        return self.compose(other)
+
+    def __call__(self, term: KInner) -> KInner:
+        return self.apply(term)
+
+    def compose(self, other: 'Subst') -> 'Subst':
+        from_other = ((k, self(v)) for k, v in other.items())
+        from_self = ((k, v) for k, v in self.items() if k not in other)
+        return Subst(dict(chain(from_other, from_self)))
```

Comment:
So in `subst1.compose(subst2)`, we are preferring the bindings from `subst2`? Seems more like an `update` than a `compose` to me.